File: platform/cli/src/commands/submit.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { execSync } from 'node:child_process';
import * as git from '../git';
import * as api from '../api';
import {
  loadConfig,
  getCurrentFeature,
  loadDiffsForFeature,
  saveDiff,
  saveFeature,
  saveUndoEntry,
} from '../workspace';
import {
  errorMessage,
  diffLabel,
  diffStatusIcon,
  featureTypePrefix,
  nextStepHint,
} from '../format';

function execQuiet(cmd: string): string | null {
  try {
    return execSync(cmd, { encoding: 'utf-8', stdio: 'pipe' }).trim();
  } catch {
    return null;
  }
}

export function registerSubmitCommand(program: Command): void {
  program
    .command('submit')
    .description('Submit diffs for review')
    .option('--preview', 'Show what would be submitted without doing it')
    .action(async (opts: { preview?: boolean }) => {
      try {
        const feature = getCurrentFeature();
        if (!feature) {
          console.log(chalk.red('No active feature.'));
          console.log(
            chalk.dim('Start one with: devpod feature "name"'),
          );
          process.exit(1);
          return;
        }

        const diffs = loadDiffsForFeature(feature);
        const draftDiffs = diffs.filter((d) => d.status === 'draft');

        if (draftDiffs.length === 0) {
          console.log(chalk.dim('All diffs already submitted.'));
          return;
        }

        // Preview mode
        if (opts.preview) {
          console.log(chalk.bold('Would submit:'));
          for (const d of draftDiffs) {
            console.log(
              `  ${diffLabel(d.position)} ${d.title} (${chalk.green(`+${d.additions}`)} ${chalk.red(`-${d.deletions}`)})`,
            );
          }
          console.log();
          console.log(
            chalk.dim(
              `Run without --preview to submit ${draftDiffs.length} diff${draftDiffs.length !== 1 ? 's' : ''}.`,
            ),
          );
          return;
        }

        // Save undo entry
        saveUndoEntry({
          action: 'submit',
          timestamp: new Date().toISOString(),
          refBefore: git.getHeadSha(),
          description: `Submit ${draftDiffs.length} diff${draftDiffs.length !== 1 ? 's' : ''}`,
          data: {
            featureSlug: feature.slug,
            diffs: draftDiffs.map((d) => d.uuid),
          },
        });

        // Push the feature branch
        git.pushForce(feature.branch);

        const config = loadConfig();

        // Create GitHub PR if none exists
        let prUrl = '';
        const existingPr = diffs.find((d) => d.githubPr !== null);

        if (!existingPr) {
          // Build PR title
          const prefix = featureTypePrefix(feature.type);
          const prTitle = `${prefix}(${feature.slug}): ${feature.name}`;

          // Build PR body from diffs
          const diffList = diffs
            .map(
              (d) =>
                `- ${diffLabel(d.position)}: ${d.title}`,
            )
            .join('\n');
          const body = `## Diffs\n\n${diffList}`;

          try {
            const result = execQuiet(
              `gh pr create --base ${config.defaultBranch} --title ${JSON.stringify(prTitle)} --body ${JSON.stringify(body)}`,
            );
            if (result) {
              prUrl = result;
            }

            // Get PR number
            const prNumStr = execQuiet(
              'gh pr view --json number --jq .number',
            );
            if (prNumStr) {
              const prNum = parseInt(prNumStr, 10);
              if (!isNaN(prNum)) {
                // Set PR number on all diffs
                for (const d of diffs) {
                  d.githubPr = prNum;
                  saveDiff(d);
                }
              }
            }
          } catch {
            // PR creation is best-effort
            console.log(
              chalk.dim('Could not create PR (gh may not be configured).'),
            );
          }
        }

        // Update draft diffs to submitted
        for (const d of draftDiffs) {
          d.status = 'submitted';
          d.updated = new Date().toISOString();
          saveDiff(d);
        }

        // Update feature status
        feature.status = 'submitted';
        saveFeature(feature);

        // Trigger CI if auto-run enabled
        if (config.ci.autoRun) {
          const isUp = await api.ping();
          if (isUp) {
            for (const d of draftDiffs) {
              try {
                await api.post('/api/runs', {
                  workflow: 'pull_request',
                  ref: feature.branch,
                  inputs: { diff: d.uuid },
                });
                d.ci = 'pending';
                saveDiff(d);
              } catch {
                // CI trigger is best-effort
              }
            }
          }
        }

        // Print summary
        console.log(
          `${chalk.green('\u2713')} Submitted ${draftDiffs.length} diff${draftDiffs.length !== 1 ? 's' : ''}`,
        );

        if (prUrl) {
          console.log(`  PR: ${chalk.cyan(prUrl)}`);
        }

        // Show stack
        const allDiffs = loadDiffsForFeature(feature);
        const stackParts = allDiffs.map((d) => {
          return `${diffLabel(d.position)} ${diffStatusIcon(d.status)}`;
        });
        console.log(chalk.dim(`  Stack: ${stackParts.join(' \u2192 ')}`));

        console.log(nextStepHint('submit'));
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}

File: platform/cli-go/internal/commands/submit.go
package commands

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
	"time"

	"github.com/briandowns/spinner"
	"github.com/elloloop/devpod/platform/cli-go/internal/api"
	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/elloloop/devpod/platform/cli-go/internal/git"
	"github.com/elloloop/devpod/platform/cli-go/internal/workspace"
	"github.com/spf13/cobra"
)

func execQuiet(name string, args ...string) string {
	out, err := exec.Command(name, args...).CombinedOutput()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func newSubmitCmd() *cobra.Command {
	var preview bool

	cmd := &cobra.Command{
		Use:   "submit",
		Short: "Push diffs for review and create/update a GitHub PR",
		Long: `Pushes the clean branch and versions branch to origin, then
creates or updates a GitHub pull request.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			// Check for pending rebase
			if err := workspace.CheckPendingRebase(); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			feature, err := workspace.ValidateOnFeatureBranch()
			if err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			diffs := workspace.LoadDiffsForFeature(*feature)
			var draftDiffs []workspace.DiffData
			for _, d := range diffs {
				if d.Status == "draft" {
					draftDiffs = append(draftDiffs, d)
				}
			}

			if len(draftDiffs) == 0 {
				fmt.Println(format.DimText("All diffs already submitted."))
				return nil
			}

			// Preview mode
			if preview {
				fmt.Println("Would submit:")
				for _, d := range draftDiffs {
					fmt.Printf("  %s %s (+%d -%d)\n",
						format.DiffLabel(d.Position), d.Title, d.Additions, d.Deletions)
				}
				fmt.Println()
				plural := "s"
				if len(draftDiffs) == 1 {
					plural = ""
				}
				fmt.Println(format.DimText(fmt.Sprintf("Run without --preview to submit %d diff%s.", len(draftDiffs), plural)))
				return nil
			}

			// Save undo entry
			headBefore, _ := git.GetHeadSHA()
			diffUUIDs := make([]string, len(draftDiffs))
			for i, d := range draftDiffs {
				diffUUIDs[i] = d.UUID
			}
			plural := "s"
			if len(draftDiffs) == 1 {
				plural = ""
			}
			_ = workspace.SaveUndoEntry(workspace.UndoEntry{
				Action:      "submit",
				Timestamp:   time.Now().UTC().Format(time.RFC3339),
				RefBefore:   headBefore,
				Description: fmt.Sprintf("Submit %d diff%s", len(draftDiffs), plural),
				Data: map[string]interface{}{
					"featureSlug": feature.Slug,
					"diffs":       diffUUIDs,
				},
			})

			s := spinner.New(spinner.CharSets[14], 100*time.Millisecond)
			s.Suffix = " Pushing branches..."
			s.Start()

			// Push the clean branch
			_ = git.PushForce(feature.Branch)

			// Push the versions branch
			vb := feature.VersionsBranch
			if vb == "" {
				vb = workspace.VersionsBranchName(feature.Branch)
			}
			if git.BranchExists(vb) {
				_ = git.PushForce(vb)
			}
			s.Stop()

			config := workspace.LoadConfig()

			// Create GitHub PR if none exists
			var prURL string
			existingPR := false
			for _, d := range diffs {
				if d.GitHubPR != 0 {
					existingPR = true
					break
				}
			}

			if !existingPR {
				prefix := format.FeatureTypePrefix(string(feature.Type))
				prTitle := fmt.Sprintf("%s(%s): %s", prefix, feature.Slug, feature.Name)

				var diffListParts []string
				for _, d := range diffs {
					vLabel := ""
					if d.Version > 1 {
						vLabel = fmt.Sprintf(" v%d", d.Version)
					}
					diffListParts = append(diffListParts, fmt.Sprintf("- %s%s: %s", format.DiffLabel(d.Position), vLabel, d.Title))
				}
				body := "## Diffs\n\n" + strings.Join(diffListParts, "\n")

				prTitleJSON, _ := json.Marshal(prTitle)
				bodyJSON, _ := json.Marshal(body)

				result := execQuiet("gh", "pr", "create",
					"--base", config.DefaultBranch,
					"--title", string(prTitleJSON),
					"--body", string(bodyJSON))
				if result != "" {
					prURL = result
				}

				// Get PR number
				prNumStr := execQuiet("gh", "pr", "view", "--json", "number", "--jq", ".number")
				if prNumStr != "" {
					prNum, err := strconv.Atoi(prNumStr)
					if err == nil {
						for i := range diffs {
							diffs[i].GitHubPR = prNum
							_ = workspace.SaveDiff(diffs[i])
						}
					}
				}
			}

			// Update draft diffs to submitted
			for i := range draftDiffs {
				draftDiffs[i].Status = "submitted"
				draftDiffs[i].Updated = time.Now().UTC().Format(time.RFC3339)
				_ = workspace.SaveDiff(draftDiffs[i])
			}

			// Update feature status
			feature.Status = "submitted"
			_ = workspace.SaveFeature(*feature)

			// Trigger CI if auto-run enabled
			if config.CI.AutoRun {
				if api.Ping() {
					for i := range draftDiffs {
						var runResult interface{}
						err := api.Post("/api/runs", map[string]interface{}{
							"workflow": "pull_request",
							"ref":      feature.Branch,
							"inputs":   map[string]string{"diff": draftDiffs[i].UUID},
						}, &runResult)
						if err == nil {
							draftDiffs[i].CI = "pending"
							_ = workspace.SaveDiff(draftDiffs[i])
						}
					}
				}
			}

			// Print summary
			fmt.Printf("%s Submitted %d diff%s\n", format.SuccessMsg("\u2713"), len(draftDiffs), plural)

			if prURL != "" {
				fmt.Printf("  PR: %s\n", prURL)
			}

			// Show stack
			allDiffs := workspace.LoadDiffsForFeature(*feature)
			sorted := sortDiffsByPosition(allDiffs)
			var stackParts []string
			for _, d := range sorted {
				stackParts = append(stackParts, fmt.Sprintf("%s %s", format.DiffLabel(d.Position), format.DiffStatusIcon(string(d.Status))))
			}
			fmt.Println(format.DimText(fmt.Sprintf("  Stack: %s", strings.Join(stackParts, " \u2192 "))))

			fmt.Println(format.NextStepHint("submit"))
			return nil
		},
	}

	cmd.Flags().BoolVar(&preview, "preview", false, "Show what would be submitted without doing it")

	return cmd
}
